core/ecschema-metadata/src/Metadata/Class.ts

```diff
@@ -705,11 +723,11 @@ export abstract class ECClass extends SchemaItem implements CustomAttributeConta
     }
 
     if (!this._mergedPropertyCache) {
-      this._mergedPropertyCache = [];
-      this.buildPropertyCacheSync(this._mergedPropertyCache, undefined);
+      this._mergedPropertyCache = new Map<string, Property>();
+      this.buildPropertyCacheSync(this._mergedPropertyCache);
```

Comment:
this is a similar problem that we recently had. Wait until the cache is completely generated and then assign the field. Doing it like this, you open a chance that an incomplete cache may be used